<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#第4章　ニューラルネットワークとKeras" data-toc-modified-id="第4章　ニューラルネットワークとKeras-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>第4章　ニューラルネットワークとKeras</a></span></li></ul></div>

# 第4章　ニューラルネットワークとKeras

In [ ]:
# 参考
import numpy as np

A = np.zeros(20)
num = np.zeros(20)

for n in range(20):
    num[n]=n/20

**Tensorboard起動**

Anaconda Prompt- Jupyter起動させる

(base) C:\Users\a4070>cd C:\Users\a4070\Python\Jupyter\PythonAI_14

(base) C:\Users\a4070\Python\Jupyter\PythonAI_14>tensorboard --logdir=logs1

***************

TensorBoard 1.11.0 at http://PC-A4070-1807-1:6006 (Press CTRL+C to quit)

***************

TensorBoard 1.11.0 at http://localhost:6006/ (Press CTRL+C to quit)

***************


** リスト4.1:データのインポート **

In [30]:
from tensorflow.python.keras.datasets import mnist


(x_train, y_train), (x_test, y_test) = mnist.load_data()

** リスト4.2:インポートしたデータの形 **

In [31]:
# インポートしたデータの形を確認
print('x_train.shape:', x_train.shape)
print('x_test.shape:', x_test.shape)
print('y_train.shape:', y_train.shape)
print('y_test.shape:', y_test.shape)

x_train.shape: (60000, 28, 28)
x_test.shape: (10000, 28, 28)
y_train.shape: (60000,)
y_test.shape: (10000,)


** リスト4.3:インポートしたデータ（特徴量）をネットワークに対応するように変形する **

In [32]:
x_train = x_train.reshape(60000, 784)
x_train = x_train/255.
x_test = x_test.reshape(10000, 784)
x_test = x_test/255.

** リスト4.4:インポートしたデータ（クラスラベル）をネットワークに対応するように変形する **

In [33]:
from tensorflow.python.keras.utils import to_categorical


y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

** リスト4.5:モデル構築の準備 **

In [34]:
from tensorflow.python.keras.models import Sequential


model = Sequential()

** リスト4.6:中間層の追加 **

In [35]:
from tensorflow.python.keras.layers import Dense


model.add(
    Dense(
        units=64, 
        input_shape=(784,),
        activation='relu'
    )
)

** リスト4.7:出力層の追加 **

In [36]:
model.add(
    Dense(
        units=10, 
        activation='softmax'
    )
)

**Tensorboard起動**

Anaconda Prompt- Jupyter起動させる

(base) C:\Users\a4070>cd C:\Users\a4070\Python\Jupyter\PythonAI_14

(base) C:\Users\a4070\Python\Jupyter\PythonAI_14>tensorboard --logdir=logs1

TensorBoard 1.11.0 at http://localhost:6006/ (Press CTRL+C to quit)

** リスト4.8:Adamを用いた、本モデルでのMNISTデータの学習**

In [39]:
from tensorflow.python.keras.callbacks import TensorBoard


model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)
tsb = TensorBoard(log_dir='./logs1')
history_adam = model.fit(
    x_train,
    y_train,
    batch_size=32,
    epochs=20,
    validation_split=0.2,
    callbacks=[tsb]
)

Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 32s 662us/step - loss: 0.0087 - acc: 0.9974 - val_loss: 0.1107 - val_acc: 0.9738
Epoch 2/20
48000/48000 [==============================] - 27s 559us/step - loss: 0.0073 - acc: 0.9977 - val_loss: 0.1256 - val_acc: 0.9721
Epoch 3/20
48000/48000 [==============================] - 26s 550us/step - loss: 0.0057 - acc: 0.9985 - val_loss: 0.1258 - val_acc: 0.9743
Epoch 4/20
48000/48000 [==============================] - 27s 553us/step - loss: 0.0073 - acc: 0.9977 - val_loss: 0.1248 - val_acc: 0.9735
Epoch 5/20
48000/48000 [==============================] - 26s 551us/step - loss: 0.0062 - acc: 0.9982 - val_loss: 0.1322 - val_acc: 0.9735
Epoch 6/20
48000/48000 [==============================] - 26s 541us/step - loss: 0.0048 - acc: 0.9987 - val_loss: 0.1212 - val_acc: 0.9750
Epoch 7/20
48000/48000 [==============================] - 25s 525us/step - loss: 0.0055 - acc: 0.9985 - val_loss: 0.

Train on 48000 samples, validate on 12000 samples

Epoch 1/20
48000/48000 [==============================] - 28s 588us/step - loss: 0.3273 - acc: 0.9070 - val_loss: 0.1884 - val_acc: 0.9459

Epoch 2/20
48000/48000 [==============================] - 26s 548us/step - loss: 0.1574 - acc: 0.9540 - val_loss: 0.1343 - val_acc: 0.9615

Epoch 3/20
48000/48000 [==============================] - 29s 598us/step - loss: 0.1130 - acc: 0.9669 - val_loss: 0.1171 - val_acc: 0.9652

Epoch 4/20
48000/48000 [==============================] - 29s 598us/step - loss: 0.0883 - acc: 0.9738 - val_loss: 0.1029 - val_acc: 0.9703

Epoch 5/20
48000/48000 [==============================] - 29s 598us/step - loss: 0.0727 - acc: 0.9784 - val_loss: 0.0973 - val_acc: 0.9702

Epoch 6/20
48000/48000 [==============================] - 30s 615us/step - loss: 0.0596 - acc: 0.9818 - val_loss: 0.0909 - val_acc: 0.9734

Epoch 7/20
48000/48000 [==============================] - 27s 557us/step - loss: 0.0500 - acc: 0.9848 - val_loss: 0.0894 - val_acc: 0.9737

Epoch 8/20
48000/48000 [==============================] - 26s 546us/step - loss: 0.0423 - acc: 0.9876 - val_loss: 0.0929 - val_acc: 0.9732

Epoch 9/20
48000/48000 [==============================] - 27s 560us/step - loss: 0.0367 - acc: 0.9887 - val_loss: 0.0886 - val_acc: 0.9740

Epoch 10/20
48000/48000 [==============================] - 27s 561us/step - loss: 0.0311 - acc: 0.9905 - val_loss: 0.1023 - val_acc: 0.9709

Epoch 11/20
48000/48000 [==============================] - 26s 537us/step - loss: 0.0274 - acc: 0.9917 - val_loss: 0.0960 - val_acc: 0.9734

Epoch 12/20
48000/48000 [==============================] - 26s 533us/step - loss: 0.0230 - acc: 0.9932 - val_loss: 0.1004 - val_acc: 0.9729

Epoch 13/20
48000/48000 [==============================] - 25s 517us/step - loss: 0.0200 - acc: 0.9943 - val_loss: 0.1080 - val_acc: 0.9705

Epoch 14/20
48000/48000 [==============================] - 26s 534us/step - loss: 0.0174 - acc: 0.9950 - val_loss: 0.1071 - val_acc: 0.9725

Epoch 15/20
48000/48000 [==============================] - 26s 545us/step - loss: 0.0162 - acc: 0.9953 - val_loss: 0.1015 - val_acc: 0.9742

Epoch 16/20
48000/48000 [==============================] - 26s 533us/step - loss: 0.0134 - acc: 0.9964 - val_loss: 0.1071 - val_acc: 0.9733

Epoch 17/20
48000/48000 [==============================] - 25s 520us/step - loss: 0.0121 - acc: 0.9968 - val_loss: 0.1122 - val_acc: 0.9740

Epoch 18/20
48000/48000 [==============================] - 25s 516us/step - loss: 0.0110 - acc: 0.9971 - val_loss: 0.1109 - val_acc: 0.9735

Epoch 19/20
48000/48000 [==============================] - 26s 540us/step - loss: 0.0083 - acc: 0.9979 - val_loss: 0.1143 - val_acc: 0.9736

Epoch 20/20
48000/48000 [==============================] - 27s 565us/step - loss: 0.0098 - acc: 0.9972 - val_loss: 0.1153 - val_acc: 0.9734

![CNT](TensorBoard_C4_0.PNG "CNT")
![CNT](TensorBoard_C4_1.PNG "CNT")
![CNT](TensorBoard_C4_2.PNG "CNT")

**Tensorboard起動**

Anaconda Prompt- Jupyter起動させる

(base) C:\Users\a4070>cd C:\Users\a4070\Python\Jupyter\PythonAI_14

(base) C:\Users\a4070\Python\Jupyter\PythonAI_14>tensorboard --logdir=logs1

TensorBoard 1.11.0 at http://localhost:6006/ (Press CTRL+C to quit)

** リスト4.9:Functional APIを用いたモデル構築のための準備**

In [44]:
from tensorflow.python.keras.datasets import mnist
from tensorflow.python.keras.utils import to_categorical
from tensorflow.python.keras.callbacks import TensorBoard
from tensorflow.python.keras.layers import Input, Dense
from tensorflow.python.keras.models import Model


(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(60000, 784)
x_train = x_train/255.
x_test = x_test.reshape(10000, 784)
x_test = x_test/255.
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)
tsb = TensorBoard(log_dir='./logs1')

** リスト4.10:Functional APIによるモデルの構築 **

In [45]:
input = Input(shape=(784, ))
middle = Dense(units=64, activation='relu')(input)
output = Dense(units=10, activation='softmax')(middle)
model = Model(inputs=[input], outputs=[output])

** リスト4.11:構築したモデルのコンパイル例 **

In [46]:
model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

** リスト4.12:MNISTのデータセットを学習する **

In [47]:
model.fit(
    x_train,
    y_train,
    batch_size=32,
    epochs=20,
    callbacks=[tsb],
    validation_split=0.2
)

Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 30s 625us/step - loss: 0.3385 - acc: 0.9047 - val_loss: 0.1955 - val_acc: 0.9438
Epoch 2/20
48000/48000 [==============================] - 29s 607us/step - loss: 0.1669 - acc: 0.9518 - val_loss: 0.1377 - val_acc: 0.9609
Epoch 3/20
48000/48000 [==============================] - 30s 628us/step - loss: 0.1213 - acc: 0.9654 - val_loss: 0.1259 - val_acc: 0.9648
Epoch 4/20
48000/48000 [==============================] - 29s 611us/step - loss: 0.0946 - acc: 0.9724 - val_loss: 0.1092 - val_acc: 0.9683
Epoch 5/20
48000/48000 [==============================] - 30s 616us/step - loss: 0.0766 - acc: 0.9775 - val_loss: 0.1003 - val_acc: 0.9695
Epoch 6/20
48000/48000 [==============================] - 28s 593us/step - loss: 0.0634 - acc: 0.9816 - val_loss: 0.1021 - val_acc: 0.9695
Epoch 7/20
48000/48000 [==============================] - 29s 597us/step - loss: 0.0527 - acc: 0.9844 - val_loss: 0.

Train on 48000 samples, validate on 12000 samples

Epoch 1/20
48000/48000 [==============================] - 8s 159us/step - loss: 0.3252 - acc: 0.9086 - val_loss: 0.1851 - val_acc: 0.9487

Epoch 2/20
48000/48000 [==============================] - 10s 202us/step - loss: 0.1560 - acc: 0.9550 - val_loss: 0.1399 - val_acc: 0.9599

Epoch 3/20
48000/48000 [==============================] - 10s 203us/step - loss: 0.1157 - acc: 0.9648 - val_loss: 0.1347 - val_acc: 0.9600

Epoch 4/20
48000/48000 [==============================] - 10s 206us/step - loss: 0.0926 - acc: 0.9722 - val_loss: 0.1204 - val_acc: 0.9645

Epoch 5/20
48000/48000 [==============================] - 9s 194us/step - loss: 0.0769 - acc: 0.9769 - val_loss: 0.1059 - val_acc: 0.9686

Epoch 6/20
48000/48000 [==============================] - 10s 209us/step - loss: 0.0628 - acc: 0.9815 - val_loss: 0.1107 - val_acc: 0.9692

Epoch 7/20
48000/48000 [==============================] - 12s 241us/step - loss: 0.0537 - acc: 0.9841 - val_loss: 0.1151 - val_acc: 0.9663

Epoch 8/20
48000/48000 [==============================] - 10s 201us/step - loss: 0.0465 - acc: 0.9862 - val_loss: 0.1028 - val_acc: 0.9701

Epoch 9/20
48000/48000 [==============================] - 11s 236us/step - loss: 0.0398 - acc: 0.9883 - val_loss: 0.1036 - val_acc: 0.9706

Epoch 10/20
48000/48000 [==============================] - 10s 205us/step - loss: 0.0345 - acc: 0.9896 - val_loss: 0.1014 - val_acc: 0.9732

Epoch 11/20
48000/48000 [==============================] - 11s 219us/step - loss: 0.0297 - acc: 0.9911 - val_loss: 0.1065 - val_acc: 0.9726

Epoch 12/20
48000/48000 [==============================] - 13s 276us/step - loss: 0.0246 - acc: 0.9929 - val_loss: 0.1054 - val_acc: 0.9719

Epoch 13/20
48000/48000 [==============================] - 11s 223us/step - loss: 0.0226 - acc: 0.9929 - val_loss: 0.1144 - val_acc: 0.9699

Epoch 14/20
48000/48000 [==============================] - 9s 197us/step - loss: 0.0187 - acc: 0.9943 - val_loss: 0.1138 - val_acc: 0.9720

Epoch 15/20
48000/48000 [==============================] - 11s 221us/step - loss: 0.0160 - acc: 0.9954 - val_loss: 0.1140 - val_acc: 0.9732

Epoch 16/20
48000/48000 [==============================] - 11s 230us/step - loss: 0.0141 - acc: 0.9962 - val_loss: 0.1211 - val_acc: 0.9714

Epoch 17/20
48000/48000 [==============================] - 10s 212us/step - loss: 0.0130 - acc: 0.9964 - val_loss: 0.1177 - val_acc: 0.9741

Epoch 18/20
48000/48000 [==============================] - 12s 245us/step - loss: 0.0113 - acc: 0.9972 - val_loss: 0.1303 - val_acc: 0.9703

Epoch 19/20
48000/48000 [==============================] - 12s 243us/step - loss: 0.0109 - acc: 0.9972 - val_loss: 0.1235 - val_acc: 0.9722

Epoch 20/20
48000/48000 [==============================] - 11s 227us/step - loss: 0.0090 - acc: 0.9977 - val_loss: 0.1256 - val_acc: 0.9742


<tensorflow.python.keras.callbacks.History at 0x18f153b8588>


![CNT](TensorBoard_C4_3.PNG "CNT")
![CNT](TensorBoard_C4_4.PNG "CNT")
![CNT](TensorBoard_C4_5.PNG "CNT")

In [1]:
from IPython.display import HTML

HTML("""
<button id="code-show-switch-btn">スクリプトを非表示にする</button>

<script>
var code_show = true;

function switch_display_setting() {
    var switch_btn = $("#code-show-switch-btn");
    if (code_show) {
        $("div.input").hide();
        code_show = false;
        switch_btn.text("スクリプトを表示する");
    }else {
        $("div.input").show();
        code_show = true;
        switch_btn.text("スクリプトを非表示にする");
    }
}

$("#code-show-switch-btn").click(switch_display_setting);
</script>
""")